[PM] - [18n] Support for i18next translations + German translation

marcelgerber opened the pull request:
This is a follow-up of #39.
[See the page live](https://googledrive.com/host/0B986Uw1nLGY7XzNlY2hXVXI5X1U/index.html?lang=de) (make sure you enable insecure content).

Changes made:
- [i18next](http://i18next.com) added. It does all the translation stuff for us, and is pretty fast.
- German translation added (`locales/de/translation.json`)
  @pthiess Could you please review the translation?
  @jadbox Could you please take a look at the structure of the translation object keys, the changes done to the `.html` files and so on?

Other changes:
- Minor changes to the english translation
- Use jQuery. Yes, using http://i18next.com without jQuery wouldn't even make sense. (We can refactor some code later to actually use jQuery in other places)
- Remove `jquery.reveal.js` - it wasn't needed and caused nothing but errors.

Commit:
Support for i18next translations, german translation

marcelgerber commented on contribute.html:
```diff
@@ -348,47 +370,85 @@ <h2 id="try-it">Videos</h2>
         if(document.documentElement.clientWidth > 480)
         {      
             // get starter issues from github
             $.ajax({url:'https://api.github.com/repos/adobe/brackets/issues?labels=starter+bug&state=open&milestone=none', dataType:'json', success:function(data){
```
I wonder why `$.ajax()` even worked before, because I can't seem to find any jQuery included...

TomMalbran replied:
jQuery or Zepto are loaded in https://github.com/adobe/brackets.io/blob/gh-pages/index.html#L338. It works in a similar way in contribute.html.

So you should load `i18next.min.js` after that line. And if it doesn't work with zepto, we need to find a way to always use jQuery.

marcelgerber commented on contribute.html:
```diff
@@ -348,47 +370,85 @@ <h2 id="try-it">Videos</h2>
+                detectLngQS: "lang", // use ?lang=... instead of ?setLng=...
+                cookieName: "lang", // use cookie "lang" instead of "i18next"
+                load: "unspecific", // only use the unspecific locale ("de" instead of "de-DE")
+                // useLocalStorage: true, // production only
```
Notice we should uncomment this line before we go into production, it enables localStorage cache for 3 days.
Same for `index.html`.

marcelgerber commented on index.html:
```diff
@@ -229,57 +238,57 @@ <h1 class="brackets-logo">
                     </li>
                     <li>
-                        <a href="http://youtu.be/kXTP8XqrSwE">Responsive Design Tool for Brackets</a>
+                        <a href="http://youtu.be/kXTP8XqrSwE" data-i18n="footer.videos.content.design-tool">Responsive Design Tool for Brackets</a>
```
@jadbox c86e5ffd13f2ebb66e9d88739750b8619d9ba646 should have removed them from `index.html` as well, right?

marcelgerber commented on index.html:
```diff
@@ -296,38 +305,38 @@ <h2 id="try-it">Videos</h2>
-                <li>200MB of available hard-disk space for installation</li>
-                <li>1280x800 display with 16-bit video card</li>
+                <li data-i18n="index.page.system-requirements.mac.list.0">Multicore Intel processor</li>
+                <li data-i18n="index.page.system-requirements.mac.list.1">Mac OS X v10.6, v10.7 or v10.8</li>
```
Should we add `Mac OS X 10.9` here?

pthiess replied:
Yes, please!

marcelgerber commented on locales/de/translation.json:
```diff
@@ -0,0 +1,210 @@
+                    },
+                    "evolving": {
+                        "header": "Es entwickelt sich weiter",
+                        "content": "Brackets ist wie ein Sandkasten, in dem neue Ideen und Werkzeuge ausprobiert werden. Neue Features und Erweiterungen kommen etwa alle 2,5 Wochen heraus. Das ist wie das ganze Jahr lang Geschenke zu bekommen."
```
@pthiess `Das ist wie das ganze Jahr Geschenke zu bekommen` just doesn't sound right to me, but I can't come up with a better solution :S
Maybe you?

pthiess replied:
@SAPlayer How about: `Das ist etwa zwanzig mal Weihnachten pro Jahr -:)`?

marcelgerber replied:
Not bad ;)

marcelgerber commented on index.html:
```diff
@@ -140,31 +149,31 @@ <h1 class="brackets-logo">
-                <h2>Linux: New &amp; Improved</h2>
-                <p>Over the last few sprints we’ve made progress on Linux (Debian/Ubuntu) support for Brackets.
+                <h2 data-i18n="index.page.sub-hero.linux.header">Linux: New &amp; Improved</h2>
+                <p><span data-i18n="index.page.sub-hero.linux.content">Over the last few releases we’ve made progress on Linux (Debian/Ubuntu) support for Brackets.</span>
```
This is one of the not-too-many changes of the english translation. `sprints` -> `releases`

marcelgerber commented:
@jadbox Notice I changed `Github` to `GitHub` and `Youtube` to `YouTube` (case-only changes).
That's the way they pronounce themselves.

TomMalbran commented:
This is great. Once is merged I can make the spanish locale.

I have 2 requests:
- Could we move the content in the script tags to a separate JavaScript file? It looks bad if we don't do this on the site.
- Could we load all the JavaScript files at the bottom of the page? Some are loaded at the top and some at the bottom.

Commit:
Rewrite thirdparty module usage + thirdparty updates

Commit:
Code review changes

marcelgerber commented:
@TomMalbran @pthiess Changes pushed.
- Thirdparty modules updated: Zepto, jQuery (minified ones are used now), Modernizr
- New modules added to Zepto (see file comment)
- Added `OS X 10.9`, changed `evolving.content`
- Loading is done in `load.i18next.js`
- `script` tags moved down
- No more `jquery.min.js` needed

marcelgerber commented:
Drive-by: We shouldn't translate the legal footer text, right?

```
Brackets was founded by Adobe as a community guided, open source project to push web development editors to the next level. 
The project is released under an MIT License.
```

Commit:
Update Foundation

marcelgerber commented:
Well, I've got a bit of a problem right now. The requirements/installation buttons won't work. It's because of the Zepto update we needed. It works in IE9 where jQuery is used (but fails in IE8 because of some weird shit).
Btw, notice that Foundation 5 won't work with Zepto (and maybe not even with jQuery 1.x, idk). http://zurb.com/article/1293/why-we-dropped-zepto

marcelgerber commented:
Notice the title isn't changed in IE8. IE9 works the way it should work.

TomMalbran commented:
@SAPlayer Great. I think that we can still clean up all the old js files into 2 specific file files. In the future we could combine all the js files.

I guess that we should just use jQuery, and since jQuery 1.11.x and 2.1.x have the same APIs any of those versions should probably work. Anyway, since we are dropping support for XP, it might make sense to also drop support for IE 6/7/8 and start using jQuery 2.x.

marcelgerber commented:
Sounds great, but I'd like to hear another ok for using just jQuery 2 (maybe @jadbox?).
Well, I hope it's fine to have this many thirdparty libs/updates in this PR. Not much than Foundation 5 left :D
Btw, I just noticed the site doesn't even work well in IE8, because there's no foreach method, but we use it to iterate through the blog posts.
It'd be nice to have a little hint stating "Your browser is no longer supported. Please update."

pthiess commented:
@SAPlayer  yep - @jadbox owns this page so I out a PM label on it and assigned it to him.

jadbox commented:
@SAPlayer Yep I'm good with using jQuery 2. Having a browser hint would be nice but not required for right now. How ready are you for this to be merged in?

marcelgerber commented:
@jadbox I won't have time until Thursday, but I can hopefully get everything done then.

Commit:
Code review changes

Commit:
Enable production mode

Add README
Add last translated SHA

marcelgerber commented:
Ready for another review. jQuery 2 is being used, README was added.
(Works in IE9, but not in IE8 because of jQuery)

pthiess commented:
@jadbox

jadbox commented:
Will review it early this week for merging!

ingorichter commented:
I'm currently reviewing the changes and had some merge conflicts because of my changes to brackets.io this morning. @SAPlayer would you mind updating the PR?

ingorichter commented on index.html:
```diff
@@ -391,17 +383,37 @@ <h2 id="try-it">Videos</h2>
+
+    <script>
+        var shortLocale = (requestedLang && requestedLang.split("-")[0]) || "en",
+            versionInfoBaseUrl = "http://dev.brackets.io/updates/stable/",
```
Please change `dev.brackets.io` to `brackets.io`. We are now loading all update notifications from brackets.io. The redirect will be gone in the near future.
And it needs to be a jsonp request. I have fixed this on master this morning and this change has to be made here too.

ingorichter commented on index.html:
```diff
@@ -421,10 +433,10 @@ <h2 id="try-it">Videos</h2>
         if (OS != "OTHER") {
             var buildNum = "38"; // SET THE BUILD NUMBER HERE
```
I filed an issue for this. That is no longer required, since we can get the information from the en.json file.

ingorichter commented on locales/README.md:
```diff
@@ -0,0 +1,23 @@
+# How to add translations for a *new* locale
+
+1. Create a subfolder of the `locales` folder whose name is the language or locale you want to
+   create a translation for. Just use its two-letter code (e.g. `en`, `de`).
```
Perhaps we should keep this open for country specific translations. We currently have pt-br and pt-pt. If the languages differ to an extent that somebody wants to provide a more specific translation, we don't have a way to support that right away. This might not be so important right now, but it could come back as a feature request.

marcelgerber replied:
The reason I didn't enable this was because i18next looks everything up, so currently we have 1 request, but with country-specific codes we would have 2 (and they, of course, take longer to load).
But we can simply enable this whenever we want. Just one line in load.i18next.js and a little README change.

ingorichter commented on locales/de/translation.json:
```diff
@@ -0,0 +1,208 @@
+                    "more": "Entdecke das Repository."
+                },
+                "setup-env": {
+                    "header": "Richte deine Entwickungsumgebung ein",
```
nit/typo: Entwicklungsumgebung

ingorichter commented:
@SAPlayer I'm done with my review and have only a couple of minor nits. Great work! And I'm happy that you removed the obsolete libs that caused errors in the console.

ingorichter commented:
@SAPlayer I forgot that ask one thing: the language will always be determined automatically. The only way to forcefully switch to a different language is to specify `?lang=locale`, correct?

Commit:
Merge

Commit:
Fix minor typo

marcelgerber commented:
Ready for another review.

Take a look at http://i18next.com/pages/doc_init.html.
1. (highest priority) UrlParam `?lang=locale`
2. (medium priority) Cookie `i18next`
3. (lowest priority) Browser language

ingorichter commented:
@SAPlayer Thanks for the update on the order to determine the locale.

ingorichter commented:
@SAPlayer Thanks for the updates. Merged.

marcelgerber commented:
Thank you. That one looks and works great!
(I may try working on Foundation 5, now that we use jQuery)
